nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -0,0 +1,135 @@
+        
+    # create the thread for receive and send data
+    th_receive = Thread(target=receive, args=(logger_receive,store,status_data,buffer, comm_receiver))
+    th_send = Thread(target=send, args=(logger_send,analyse,status_data,buffer, comm_sender))
+
+    # start the threads
+    # FAT END POINT
+    logger_master.info('Start thread')
+    th_receive.start()
+    th_send.start()
+    th_receive.join()
+    th_send.join()
+    logger_master.info('thread join')
+    return
+    
+    
+def receive(logger,store,status_data,buffer, comm):
+    '''
+    the receive part of the translator
+    :param logger : logger fro the thread
+    :param store : object for store the data before analysis
+    :param status_data: the status of the buffer (SHARED between thread)
+    :param buffer: the buffer which contains the data (SHARED between thread)
+    :return:
+    '''
+    # initialise variables for the loop
+    status_ = MPI.Status() # status of the different message
+    source_sending = np.arange(0,comm.Get_remote_size(),1) # list of all the process for the commmunication
+    check = np.empty(1,dtype='b')
+    count=0
+    while True: # FAT END POINT
+        # send the confirmation of the process can send data
+        logger.info(" Nest to TVB : wait all")
+        for source in source_sending:
+            comm.Recv([check, 1, MPI.CXX_BOOL], source=source, tag=MPI.ANY_TAG, status=status_)
+
+        if status_.Get_tag() == 0:
+            logger.info(" Nest to TVB : start to receive")
+            #  Get the data/ spike
+            for source in source_sending:
+                comm.Send([np.array(True,dtype='b'),MPI.BOOL],dest=source,tag=0)
+                shape = np.empty(1, dtype='i')
+                comm.Recv([shape, 1, MPI.INT], source=source, tag=0, status=status_)
+                data = np.empty(shape[0], dtype='d')
+                comm.Recv([data, shape[0], MPI.DOUBLE], source=source, tag=0, status=status_)
+                store.add_spikes(count,data)
+            while status_data[0] != 1 and status_data[0] != 2: # FAT END POINT
+                time.sleep(0.001)
+                pass
+            # wait until the data can be send to the sender thread
+            # Set lock to true and put the data in the shared buffer
+            buffer[0] = store.return_data()
+            with lock_status: # FAT END POINT
+                if status_data[0] != 2:
+                    status_data[0] = 0
+        elif status_.Get_tag() == 1:
+            logger.info("Nest to TVB : receive end " + str(count))
+            count += 1
+        elif status_.Get_tag() == 2:
+            with lock_status:
+                status_data[0] = 2
+            logger.info("end simulation")
+            break
+        else:
+            raise Exception("bad mpi tag"+str(status_.Get_tag()))
+    logger.info('communication disconnect')
+    comm.Disconnect()
+    logger.info('end thread')
+    return
```

Comment:
Spurious return

Reply:
Old code of the threaded version. Will be changed.